### PR TITLE
Linkit / EV3 howto

### PR DESCRIPTION
**Pull request by cavedunissin**
update LinkIt / Lego EV3 tutorials, code and HowTos entries. (Hope there will no errors) - David.

### CODE REVIEW

**Comment by ewpatton** [comment]
Thanks for the pull request, David. GitHub is reporting that there are two convicts with master. Would you please either merge master into this branch or rebase this on top of master to resolve the conflicts?

File: docs/makepages.py [hunk 1]
```diff
@@ -100,7 +100,14 @@
       <Route path="arduinoservo" component={Arduino101ServoPage} />
       <Route path="arduinosoundrecorder" component={Arduino101SoundRecorderPage} />
     </Route>
-    <Route path="faq">
+    <Route path="linkit">
+	  <Route path="linkitintro" component={linkitintroPage} />
```
Comment (ewpatton) [inline docs/makepages.py]: After compiling, the page doesn't load here due to capitalization mismatch between the `...Page` components and the file names in the pages directory.

File: docs/src/app/HowTos.js [hunk 2]
```diff
@@ -142,10 +142,42 @@ const howtosList = [
...
     img: 'assets/howtos/LinkIt_7697.png',
-    link: 'assets/tutorials/MIT_App_Inventor_LED_7697.pdf',
+    link: 'assets/tutorials/MIT_App_Inventor_7697_LED.pdf',
     description: 'Control the LED light on LinkIt 7697 (Arduino compatible). Turn the light on and off by buttons and speech recognizer.',
     equipment: ['LinkIt 7697', 'LED (optional)'],
     time: 20,
+  },{
+    title: 'How to: Read analog data',
+    author: 'MIT App Inventor Project',
+    img: 'assets/howtos/LinkIt_7697.png',
```
Comment (ewpatton) [inline docs/src/app/HowTos.js]: Verify these image files. They don't show up if I fix the errors above.

File: docs/src/app/pages/LinkIt7688.md [hunk 3]
```diff
@@ -0,0 +1,27 @@
+# Platform: Arduino 101
```
Comment (ewpatton) [inline docs/src/app/pages/LinkIt7688.md]: This file is just a copy of the Arduino 101 page. Please update it before we can finish reviewing.

File: docs/src/app/pages/LinkIt7697.md [hunk 4]
```diff
@@ -0,0 +1,27 @@
+# Platform: Arduino 101
```
Comment (ewpatton) [inline docs/src/app/pages/LinkIt7697.md]: Ditto this being a copy.

File: docs/src/app/pages/LinkItintro.md [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+![Image of the LinkIt 7697. Source: arduino.cc](/assets/howtos/LinkIt7697.png)
+<br>
+Image source: <a href="https://www.arduino.cc/en/Main/ArduinoBoard101" target="_blank">Arduino 101 page on Arduino.cc</a>
+</div>
+
+MIT App Inventor provides a sketch for the Arduino 101 and extensions to interact with the sketch. Please follow [these instructions](/assets/tutorials/MIT_App_Inventor_IoT_Setup.pdf) to program your Arduino 101 with the [App Inventor sketch](/assets/resources/AIM-for-Things-Arduino101.zip).
+
+The following extensions are available for use with the companion sketch as part of the [Arduino 101 Extension Bundle](/assets/resources/edu.mit.appinventor.iot.arduino101.aix):
+
+* [Arduino 101 Accelerometer](#/arduino101/arduinoaccelerometer) &mdash; Access the onboard digital 3-axis accelerometer on the Curie's inertial measurement unit (IMU) chip.
```
Comment (ewpatton) [inline docs/src/app/pages/LinkItintro.md]: Do all of the Arduino 101 extensions work with the LinkIt? Otherwise you should update this list.

**Review by ewpatton — CHANGES_REQUESTED** [review]
Hi David. Please see the feedback I left. At the moment if I cleanly check out this branch and try to run it there are issues w.r.t. file name case and image paths.

**Review by ewpatton — CHANGES_REQUESTED** [review]
@cavedunissin You should also update the top navigation and the devices page to include entries for the LEGO and LinkIt. I would also suggest putting the LEGO entry before LinkIt to keep the list sorted alphabetically.

**Comment by ewpatton** [comment]
Take a look at Master.js (to handle page titles), DocumentationMenu.js (to add to top menu), and DeviceList.js (to add to device table).

File: docs/src/app/HowTos.js [hunk 6]
```diff
@@ -141,12 +141,71 @@ const howtosList = [
...
+  }, {
+    title: 'How to: RGB LED control (Wi-Fi)',
+    author: 'MIT App Inventor Project',
+    img: 'assets/howtos/RGBLED.jpg',
+    link: 'assets/howtos/MIT_App_Inventor_7697_WIFILED.pdf',
+    description: 'Control LinkIt 7697 RGB LED to shine in four different colors through Wi-Fi',
+    equipment: ['LinkIt 7697', 'RGB LED'],
+    time: 90,
+  },
+  /*{
```
Comment (ewpatton) [inline docs/src/app/HowTos.js]: I'd prefer you not include commented out entries. Add them when they are ready.

File: docs/src/app/HowTos.js [hunk 7]
```diff
@@ -141,12 +141,71 @@ const howtosList = [
...
+  },{
+    title: 'How to: Sensor Panel',
+    author: 'MIT App Inventor Project',
+    img: 'assets/howtos/LegoEV3.png',
+    link: 'assets/howtos/MIT_App_Inventor_EV3_SensorPanel.pdf',
+    description: 'View Lego EV3 sensor status on app screen: touch, ultrasonic and color sensors.',
+    equipment: ['Lego EV3 robot kit'],
+    time: 60,
+  },
+  /*{
```
Comment (ewpatton) [inline docs/src/app/HowTos.js]: See comment above about commented-out code.

File: docs/src/app/Master.js [hunk 8]
```diff
@@ -211,6 +212,7 @@ class Master extends Component {
               <ForStudentsMenu />
               <ForMakersMenu />
               <DocumentationMenu />
+			  <haha />
```
Comment (ewpatton) [inline docs/src/app/Master.js]: What is this?

File: docs/src/app/pages/LinkIt7697.md [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+* Peripheral interfaces including GPIO, UART, I2C, SPI, PWM, EINT, ADC, IrDA, I2S
+* Pin headers for SWD debugging
+* LinkIt SDK GCC, IAR Embedded Workbench, and Keil uVision IDE
+* Additional add-on Package for Arduino IDE support
+* Size: 48 x 26 mm
+* Open hardware board reference design including schematic, <a href="https://labs.mediatek.com/en/download/1ega2lbl" target="_blank">layouts and pin details.</a>
+
+Please follow our instructions:
+
+* [LED control (BluetoothLE)](/assets/howtos/MIT_App_Inventor_7697_LED.pdf) &mdash; Control the onboard LED by button and voice through Bluetooth.
```
Comment (ewpatton) [inline docs/src/app/pages/LinkIt7697.md]: The tutorials for Arduino and micro:bit open in a new tab. You should look at those files for an example so that the behavior is the same.

File: docs/src/app/pages/legoev3intro.md [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+With App Inventor's LegoEv3 components, you can interact Lego EV3 robots easily.
+
+<div style="text-align: center; font-size: 75%; margin: 16pt 0;">
+![Image of the Arduino/Genuino 101 board. Source: Lego.com](/assets/howtos/LegoEV3.png)
+<br>
+Image source: <a href="https://www.lego.com/en-us/mindstorms/about-ev3" target="_blank">Lego EV3 electrical components on Lego.com</a>
+</div>
+
+MIT App Inventor provides complete support to interact with Lego EV3, including:
+* [Ev3Motors](http://ai2.appinventor.mit.edu/reference/components/legomindstorms.html#Ev3Motors) &mdash; control motor speed and direction.
```
Comment (ewpatton) [inline docs/src/app/pages/legoev3intro.md]: Ditto above about opening external links in a new tab.

File: docs/src/app/pages/legoev3intro.md [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+* [Ev3GyroSensor](http://ai2.appinventor.mit.edu/reference/components/legomindstorms.html#Ev3GyroSensor) &mdash; detect angular velocity.
+* [Ev3TouchSensor](http://ai2.appinventor.mit.edu/reference/components/legomindstorms.html#Ev3TouchSensor) &mdash; detect whether this sensor is pressed/released.
+* [Ev3UltrasonicSensor](http://ai2.appinventor.mit.edu/reference/components/legomindstorms.html#Ev3UltrasonicSensor) &mdash; detect distance between object ahead.
+* [Ev3UI](http://ai2.appinventor.mit.edu/reference/components/legomindstorms.html#Ev3UI) &mdash; to diplay text/number/image on Ev3's 100x64 pixel monochrome LED.
+* [Ev3Sound](http://ai2.appinventor.mit.edu/reference/components/legomindstorms.html#Ev3Sound) &mdash; to play sounds or shart melody on Ev3 brick.
+* [Ev3Commands](http://ai2.appinventor.mit.edu/reference/components/legomindstorms.html#Ev3Commands) &mdash; low level commands can talk to Ev3 brick directly.
+
+Please check tutorials:
+
+* [Button Control](assets/howtos/MIT_App_Inventor_EV3_ButtonControl.pdf) &mdash; Control Lego EV3 robot with buttons. You can also adjust the motor speed by sliders..
```
Comment (ewpatton) [inline docs/src/app/pages/legoev3intro.md]: Open link in a new tab.

**Review by ewpatton — CHANGES_REQUESTED** [review]
@cavedunissin Almost there. Please address the comments I made and then we can put this up on the test server.

**Comment by ewpatton** [comment]
@cavedunissin You should also merge the master branch from mit-cml into your branch and resolve the conflict on the Magnetometer PDF file.

File: docs/src/app/pages/LinkIt7697.md [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+* Two buttons for system reset and external interrupt
+* Peripheral interfaces including GPIO, UART, I2C, SPI, PWM, EINT, ADC, IrDA, I2S
+* Pin headers for SWD debugging
+* LinkIt SDK GCC, IAR Embedded Workbench, and Keil uVision IDE
+* Additional add-on Package for Arduino IDE support
+* Size: 48 x 26 mm
+* Open hardware board reference design including schematic, <a href="https://labs.mediatek.com/en/download/1ega2lbl" target="_blank">layouts and pin details.</a>
+
+Please follow our instructions:
+<ul><li><a href='http://iot.appinventor.mit.edu/assets/howtos/MIT_App_Inventor_7697_LED.pdf' target='_blank'>LED Control (BluetoothLE)</a> &mdash; Control the onboard LED by button and voice through Bluetooth.</li></ul>
```
Comment (ewpatton) [inline docs/src/app/pages/LinkIt7697.md]: Rather than providing an absolute URL, use a relative URL starting with `/`, for example `/assets/howtos/MIT_App_Inventor_7697_LED.pdf`. Right now all of these links are broken when testing locally because the URLs point to the production site.

Comment (ewpatton) [inline docs/src/app/pages/LinkIt7697.md]: Rather than having a `<ul>` per line, you should start with a `<ul>` and only put the `</ul>` at the end of the list. As given, you are creating a new list per line rather than the more desirable multiple `<li>` in a single list.

File: docs/src/app/pages/MicroBitAccelerometer.md [hunk 13]
```diff
@@ -4,6 +4,10 @@ The <code>Microbit_Accelerometer</code> component lets users configure the BBC m
 
 The accelerometer's report rate, or period, determines how frequently data will be sent to App Inventor.
 
+<strong>More links:</strong><ul><li>Download a <a href='http://iot.appinventor.mit.edu/assets/samples/MicrobitAccelerometer.aia' target='_blank'>sample projects</a> for the micro:bit onboard buttons.</li>
+                                <li>View the <a href='http://iot.appinventor.mit.edu/assets/howtos/MIT_App_Inventor_Microbit_Accelerometer.pdf' target='_blank'>how to instructions</a> for the micro:bit onboard buttons</li></ul>
```
Comment (ewpatton) [inline docs/src/app/pages/MicroBitAccelerometer.md]: Remove the extraneous whitespace at the start of this line.

File: docs/src/app/pages/MicroBitButton.md [hunk 14]
```diff
@@ -3,6 +3,9 @@
 The <code>Microbit_Button</code> extension provides App Inventor with information about the state of the BBC micro:bit's buttons. Developers can use this extension to request updates for when a user presses a button or read the current state of the buttons. The buttons have three states:<br>
 <ul><li>0 - up</li><li>1 - down</li><li>2 - long pressed</li></ul>
 
+<strong>More links:</strong><ul><li>Download a <a href='http://iot.appinventor.mit.edu/assets/samples/MicrobitButton.aia' target='_blank'>sample projects</a> for the micro:bit onboard buttons.</li>
+                                <li>View the <a href='http://iot.appinventor.mit.edu/assets/howtos/MIT_App_Inventor_Microbit_Button.pdf' target='_blank'>how to instructions</a> for the micro:bit onboard buttons</li></ul>
```
Comment (ewpatton) [inline docs/src/app/pages/MicroBitButton.md]: Ditto whitespace.

File: docs/src/app/pages/MicroBitDeviceInformation.md [hunk 15]
```diff
@@ -1,6 +1,10 @@
 # Microbit\_Device\_Information
 
-Component for Microbit_Device_Information
+The <code>Microbit_Device_Information</code> extension provides App Inventor with information about Microbit, including Firmware version, Manufacturer name and serial nuber, etc.
+
+<strong>More links:</strong><ul><li>Download a <a href='http://iot.appinventor.mit.edu/assets/samples/MicrobitDeviceInformation.aia' target='_blank'>sample projects</a> for the micro:bit onboard buttons.</li>
+                                <li>View the <a href='http://iot.appinventor.mit.edu/assets/howtos/MIT_App_Inventor_Microbit_DeviceInformation.pdf' target='_blank'>how to instructions</a> for the micro:bit onboard buttons</li></ul>
```
Comment (ewpatton) [inline docs/src/app/pages/MicroBitDeviceInformation.md]: Ditto whitespace.

File: docs/src/app/pages/MicroBitIoPin.md [hunk 16]
```diff
@@ -1,7 +1,7 @@
 # Microbit\_Io\_Pin
 
 The <code>Microbit_Io_Pin</code> component lets users configure the BBC micro:bit's analog pins for input and output, and to read, write, and request notifications for the I/O pin states.
-
+
```
Comment (ewpatton) [inline docs/src/app/pages/MicroBitIoPin.md]: Random space added to this line should be removed.

File: docs/src/app/pages/MicroBitMagnetometer.md [hunk 17]
```diff
@@ -2,6 +2,9 @@
 
 The <code>Microbit_Magnetometer</code> provides the ability to configure the BBC micro:bit's on-board magnetometer and receive one or more magnetometer samples via the appropriate methods.
 
+<strong>More links:</strong><ul><li>Download a <a href='http://iot.appinventor.mit.edu/assets/samples/MicrobitMagnetometer.aia' target='_blank'>sample projects</a> for the micro:bit onboard magnetometer.</li>
+                                <li>View the <a href='http://iot.appinventor.mit.edu/assets/howtos/MIT_App_Inventor_Microbit_Magnetometer.pdf' target='_blank'>how to instructions</a> for the micro:bit onboard magnetometer</li></ul>
```
Comment (ewpatton) [inline docs/src/app/pages/MicroBitMagnetometer.md]: Ditto whitespace.

File: docs/src/app/pages/legoev3intro.md [hunk 18]
```diff
@@ -0,0 +1,25 @@
...
+With App Inventor's LegoEv3 components, you can interact Lego EV3 robots easily.
+
+<div style="text-align: center; font-size: 75%; margin: 16pt 0;">
+![Image of the Arduino/Genuino 101 board. Source: Lego.com](/assets/howtos/LegoEV3.png)
+<br>
+Image source: <a href="https://www.lego.com/en-us/mindstorms/about-ev3" target="_blank">Lego EV3 electrical components on Lego.com</a>
+</div>
+
+MIT App Inventor provides complete support to interact with Lego EV3, including:
+<ul><li><a href='http://ai2.appinventor.mit.edu/reference/components/legomindstorms.html#Ev3Motors' target='_blank'>Ev3Motors</a> &mdash; control motor speed and direction.</li></ul>
```
Comment (ewpatton) [inline docs/src/app/pages/legoev3intro.md]: See my previous note about `<ul>`, `</ul>`, and `<li>` for this list.

File: docs/src/app/pages/legoev3intro.md [hunk 19]
```diff
@@ -0,0 +1,25 @@
...
+<ul><li><a href='http://ai2.appinventor.mit.edu/reference/components/legomindstorms.html#Ev3UltrasonicSensor' target='_blank'>Ev3UltrasonicSensor</a> &mdash; detect distance between object ahead.</li></ul>
+<ul><li><a href='http://ai2.appinventor.mit.edu/reference/components/legomindstorms.html#Ev3UI' target='_blank'>Ev3UI</a> &mdash; diplay text/number/image on Ev3's 100x64 pixel monochrome LED.</li></ul>
+<ul><li><a href='http://ai2.appinventor.mit.edu/reference/components/legomindstorms.html#Ev3Sound' target='_blank'>Ev3Sound</a> &mdash; play music or short sound effects on Ev3 brick.</li></ul>
+<ul><li><a href='http://ai2.appinventor.mit.edu/reference/components/legomindstorms.html#Ev3Commands' target='_blank'>Ev3Commands</a> &mdash; low level commands can talk to Ev3 brick directly.</li></ul>
+
+Please check Lego EV3 tutorials:
+<ul><li><a href='http://iot.appinventor.mit.edu/assets/howtos/MIT_App_Inventor_EV3_ButtonControl.pdf' target='_blank'>Button Control</a> &mdash; Control Lego EV3 robot with buttons. You can also adjust the motor speed by sliders.</li></ul>
```
Comment (ewpatton) [inline docs/src/app/pages/legoev3intro.md]: See my previous note about relative vs absolute URLs and about `<ul>`, `</ul>`, and `<li>`.

File: docs/src/www/assets/samples/WIFI_7697_wheelrobot.ino.ino [hunk 20]
```diff
@@ -0,0 +1,247 @@
+/*
```
Comment (ewpatton) [inline docs/src/www/assets/samples/WIFI_7697_wheelrobot.ino.ino]: This file has two .ino extensions. It might be confusing to people, so you should rename the file with only a single .ino.

**Review by ewpatton — DISMISSED** [review]
@cavedunissin Please check my comments about the changes you made in response to the last review. In particular be aware of the semantics of `<ul>` and also try to use relative URLs for the IOT site where possible otherwise it makes testing difficult.

File: docs/src/app/pages/MicroBitIoPin.md [hunk 21]
```diff
@@ -1,6 +1,6 @@
 # Microbit\_Io\_Pin
 
-The <code>Microbit_Io_Pin</code> component lets users configure the BBC micro:bit's analog pins for input and output, and to read, write, and request notifications for the I/O pin states.
+The <code>Microbit_Io_Pin</code> component lets users configure the BBC micro:bit's analog pins for input and output, and to read, write, and request notifications for the I/O pin states.<br>
```
Comment (ewpatton) [inline docs/src/app/pages/MicroBitIoPin.md]: Why did you add a `<br>` here where there wasn't one previously?

**Review by ewpatton — COMMENTED** [review]
See comment

File: docs/src/app/pages/MicroBitIoPin.md (hunk 21 above)
Comment (cavedunissin) [reply docs/src/app/pages/MicroBitIoPin.md]: Evan, I refer to MicroBitLed.md  to compare the difference

**Comment by ewpatton** [comment]
When I am on campus tomorrow I'll push this up to the test server so that we can have others try it out.

**Comment by ewpatton** [comment]
@cavedunissin I think it would be good to also include the Codi Bot in the devices table. It's only listed in the text at the moment.

File: docs/src/app/pages/codibotintro.md [hunk 22]
```diff
@@ -0,0 +1,35 @@
+# Platform: MIT App Inventor Codi Bot
+
+The MIT App Inventor Codi Bot is a hands-on IoT kit. Our adorable mascot can be controlled via App Inventor through Bluetooth communication. This educational kit is easy to use and guides users through the whole process of development, from building a robot to programming it. 
+
+<div style="text-align: center; font-size: 75%; margin: 16pt 0;">
+![Image of the MIT App Inventor Codi Bot.](/assets/howtos/codibot1.jpg)
+
+</div>
+
+## Parts list (<a href="https://www.robotkingdom.com.tw/product/mit-app-inventor-codi-bot/" target="_blank">Purchase link)</a>:
```
Comment (ewpatton) [inline docs/src/app/pages/codibotintro.md]: Move the closing parenthesis outside of the `</a>`.

File: docs/src/app/pages/codibotintro.md [hunk 23]
```diff
@@ -0,0 +1,35 @@
...
+* <a href="https://www.robotkingdom.com.tw/product/linkit-7697/" target="_blank">LinkIt 7697 MCU board</a>, 1
+* <a href="https://www.robotkingdom.com.tw/product/linkit-7697%E6%A9%9F%E5%99%A8%E4%BA%BA%E6%93%B4%E5%85%85%E6%9D%BF-robot-shield-for-linkit-7697/" target="_blank">Robot Shield for LinkIt 7697 V2.0</a>, 1 (with an onboard buzzer) 
+* RGB LED (common cathode), 1
```
Comment (ewpatton) [inline docs/src/app/pages/codibotintro.md]: Are we intending to link these other components? It seems odd to me that only the first two are linked.

File: docs/src/app/pages/codibotintro.md [hunk 24]
```diff
@@ -0,0 +1,35 @@
...
+* Acrylic pieces (base, wings), 2 pieces.
+* Cow leather pieces (stripes and body cusions), 4 pieces.
+* 3D printed parts, 9 pieces.
+
+## Please follow our tutorials to get started:
+
+Compared with [LinkIt 7697 tutorials](#/linkit/linkit7697), tutorials for the MIT App Inventor Codi Bot use an extension of the LinkIt 7697 dev board instead of low-level Bluetooth commands.
+
+<ul>
+<li><a href='/assets/howtos/MIT_App_Inventor_CodiBot_UserManual.pdf' target='_blank'>User manual</a> &mdash; Follow this manual to assemble the MIT App Inventor Codi Bot.</li>
```
Comment (ewpatton) [inline docs/src/app/pages/codibotintro.md]: Has Hedge reviewed the PDFs for spelling and grammar?

File: docs/src/app/pages/codibotintro.md [hunk 25]
```diff
@@ -0,0 +1,35 @@
...
+* mini servo (SG90), 2
+* female-female jump wires, 10
+* screws (M3x8, 7pcs; M2x12, 2pcs; M2x6, 12pcs)
+* Acrylic pieces (base, wings), 2 pieces.
+* Cow leather pieces (stripes and body cusions), 4 pieces.
+* 3D printed parts, 9 pieces.
+
+## Please follow our tutorials to get started:
+
+Compared with [LinkIt 7697 tutorials](#/linkit/linkit7697), tutorials for the MIT App Inventor Codi Bot use an extension of the LinkIt 7697 dev board instead of low-level Bluetooth commands.
```
Comment (ewpatton) [inline docs/src/app/pages/codibotintro.md]: I'm not sure that we need to justify/explain this here. I think the heading should be "Instructions and Samples" and then you can in the paragraph say "Once you have received your kit, please follow our tutorials to get started."

File: docs/src/app/pages/codibotintro.md (hunk 24 above)
Comment (ewpatton) [reply docs/src/app/pages/codibotintro.md]: Also, is there a Google Doc somewhere, or something similar? I have some comments on the PDFs as well that I'd like to share.

File: docs/src/www/assets/samples/BLE_7697_BLE_WalkingRobot.ino [hunk 26]
```diff
@@ -0,0 +1,201 @@
+/* To control OTTO's wlkaing and light with MIT App Inventor BLE and turn */
```
Comment (ewpatton) [inline docs/src/www/assets/samples/BLE_7697_BLE_WalkingRobot.ino]: OTTO => Codi

**Review by ewpatton — CHANGES_REQUESTED** [review]
See the comments I left. Also, please remove the Thumbs.db files and add it to your .gitignore.

File: docs/src/app/pages/codibotintro.md (hunk 22 above)
Comment (cavedunissin) [reply docs/src/app/pages/codibotintro.md]: ok

File: docs/src/app/pages/codibotintro.md (hunk 23 above)
Comment (cavedunissin) [reply docs/src/app/pages/codibotintro.md]: Cuz the first two are the core component of Codi Bot, other are quite common and easy to get. You suggest I link all of them or remove the link of first two?

File: docs/src/app/pages/codibotintro.md (hunk 24 above)
Comment (cavedunissin) [reply docs/src/app/pages/codibotintro.md]: no problem. mail to you

**Review by ewpatton — APPROVED** [review]
